integration_tests/src/bin/benchmark/main.rs

```diff
@@ -233,98 +294,248 @@ fn run_single_test(ctx: GraphicsContext, bench_config: SingleBenchConfig) -> boo
                     raw_options: Vec::new(),
                 })),
             },
         },
-    );
+    )?))
+}
 
-    let Ok(receiver) = receiver_result else {
-        return false;
-    };
+fn register_pipeline_raw_output(
+    pipeline: &Arc<Mutex<Pipeline>>,
+    output_id: OutputId,
+    output_video_options: OutputVideoOptions,
+    bench_config: &SingleBenchConfig,
+) -> Result<Box<dyn PipelineReceiver + Send>, Box<dyn std::error::Error>> {
+    let x = Pipeline::register_raw_data_output(
+        pipeline,
+        output_id,
+        RegisterOutputOptions {
+            video: Some(output_video_options),
 
-    Pipeline::start(&pipeline);
+            audio: None,
+            output_options: RawDataOutputOptions {
+                audio: None,
+                video: Some(RawVideoOptions {
+                    resolution: Resolution {
+                        width: bench_config.output_resolution.width,
+                        height: bench_config.output_resolution.height,
+                    },
+                }),
+            },
+        },
+    )?;
+    Ok(Box::new(x.video.unwrap()))
+}
 
-    let start_time = Instant::now();
-    while Instant::now() - start_time < bench_config.warm_up_time {
-        _ = receiver.recv().unwrap();
-    }
+fn register_pipeline_mp4_input(
+    pipeline: &Arc<Mutex<Pipeline>>,
+    input_id: InputId,
+    bench_config: &SingleBenchConfig,
+) -> Result<InputInitInfo, RegisterInputError> {
+    let video_decoder = bench_config.video_decoder;
+    Pipeline::register_input(
+        pipeline,
+        input_id,
+        RegisterInputOptions {
+            input_options: InputOptions::Mp4(Mp4Options {
+                should_loop: true,
+                video_decoder,
+                source: Source::File(bench_config.file_path.clone()),
+            }),
+            queue_options: QueueInputOptions {
+                offset: Some(Duration::ZERO),
+                required: true,
+                buffer_duration: None,
+            },
+        },
+    )
+}
 
-    let start_time = Instant::now();
-    let mut produced_frames: usize = 0;
-    while start_time.elapsed() < bench_config.measured_time {
-        _ = receiver.recv().unwrap();
-        produced_frames += 1;
-    }
+fn register_pipeline_raw_input(
+    pipeline: &Arc<Mutex<Pipeline>>,
+    input_id: InputId,
+) -> Result<Sender<PipelineEvent<Frame>>, RegisterInputError> {
+    let x = Pipeline::register_raw_data_input(
+        pipeline,
+        input_id,
+        RawDataInputOptions {
+            video: true,
+            audio: false,
+        },
+        QueueInputOptions {
+            offset: Some(Duration::ZERO),
+            required: true,
+            buffer_duration: None,
+        },
+    )?;
+    Ok(x.video.unwrap())
+}
 
-    let end_time = Instant::now();
+fn raw_data_sender(bench_config: SingleBenchConfig, senders: Vec<Sender<PipelineEvent<Frame>>>) {
+    let mut file = File::open(bench_config.file_path).unwrap();
```

Comment:
I get requiring h264, but how do I get this file? Benchamarks should be self contained if possible.

Also, reading from the disk is not great, I get that it is only one read per input, but why it is necessary? Why not just generate a few static images (allocate Bytes once) and use them interchangeably to build FrameData?